### PR TITLE
Fully automatic weaponry implementation [semi, burst, automatic]

### PR DESCRIPTION
**Pull request by Gandalf2k15**
## About The Pull Request

So our overlords implemented a new system for ballistic weaponry, REAL automatic guns. HOWEVER, it has one big caveat, this only applies to the laser gatling gun and the L6 Saw... so me being me, implemented a system to enable ANY gun to have either burst fire, semi-auto or auto fire modes. Using an "intelligent" system to determine what gun gets what.

New gun behaviours to take note of: 
All guns have a selector action type, pressing this will cycle through the available firemodes.

Example of use:
https://streamable.com/h1r78a

image

What this really means is that most guns you'd consider to be fully auto(assault rifles) will have the 3 modes.

## Changelog
:cl:
add: Fully implemented a gun select fire mode system. You'll notice most guns now come with a full set of fire options, semi, burst and auto. Depending on what gun you have.
/:cl:

### CODE REVIEW

**Comment by NullDagaf** [comment]
Jesus fuck, how hard it is to make a commit not error?

..don't look at mine.

**Comment by Gandalf2k15** [comment]
> Jesus fuck, how hard it is to make a commit not error?
> 
> ..don't look at mine.

The error is just monke, nothing to worry about.

File: code/modules/projectiles/gun.dm [hunk 1]
```diff
@@ -86,6 +109,29 @@
...
+
+	if(fire_select_modes.len > 1)
+		firemode_action = new(src)
+		firemode_action.button_icon_state = "fireselect_[fire_select]"
+		firemode_action.UpdateButtonIcon()
+
+/obj/item/gun/ComponentInitialize()
+	. = ..()
+	if(SELECT_FULLY_AUTOMATIC in fire_select_modes)
+		AddComponent(/datum/component/automatic_fire, fire_delay)
```
Comment (Gandalf2k15) [inline code/modules/projectiles/gun.dm]:
```suggestion
		//AddComponent(/datum/component/automatic_fire, fire_delay)
```

File: code/modules/projectiles/gun.dm [hunk 2]
```diff
@@ -86,6 +109,29 @@
...
+	sortList(fire_select_modes, /proc/cmp_numeric_asc)
+
+	if(fire_select_modes.len > 1)
+		firemode_action = new(src)
+		firemode_action.button_icon_state = "fireselect_[fire_select]"
+		firemode_action.UpdateButtonIcon()
+
+/obj/item/gun/ComponentInitialize()
+	. = ..()
+	if(SELECT_FULLY_AUTOMATIC in fire_select_modes)
```
Comment (Gandalf2k15) [inline code/modules/projectiles/gun.dm]:
```suggestion
	//if(SELECT_FULLY_AUTOMATIC in fire_select_modes)
```

**Comment by Gandalf2k15** [comment]
FINALLY SHITECODE BEGONE

**Comment by Gandalf2k15** [comment]
Merging due to prospective merge conflicts.

New component system confirmed to work via testing.

**Pull request merged**